### PR TITLE
Surface fitting updates

### PR DESCRIPTION
**Pull request by kmittal2**
Misc updates related to surface fitting capability.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4116](https://github.com/mfem/mfem/pull/4116) | @kmittal2 | @tzanio | @vladotomov + @tzanio | 2/13/24 | ⌛due 2/27/24 | ⌛due 3/5/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/tmop.cpp [hunk 1]
```diff
@@ -3004,33 +3012,180 @@ void TMOP_Integrator::EnableSurfaceFitting(const ParGridFunction &s0,
    surf_fit_eval->SetInitialField
    (*surf_fit_gf->FESpace()->GetMesh()->GetNodes(), *surf_fit_gf);
    surf_fit_gf_bg = false;
+
+   // Unify marker array across processor boundaries
```
Comment (vladotomov) [inline fem/tmop.cpp:3016]: Why would they be different?
Shouldn't this be done by the caller?

File: fem/tmop.cpp [hunk 2]
```diff
@@ -3004,33 +3012,180 @@ void TMOP_Integrator::EnableSurfaceFitting(const ParGridFunction &s0,
...
+
+   const H1_FECollection *fec = dynamic_cast<const H1_FECollection *>
+                                (surf_fit_gf->FESpace()->FEColl());
+   if (!fec)  { return; }
+   H1_FECollection *fec_grad = new H1_FECollection(fec->GetOrder(),
+                                                   dim,
+                                                   fec->GetBasisType());
+   ParFiniteElementSpace *fes_grad = new ParFiniteElementSpace
+   (pmesh,
+    fec_grad,
```
Comment (vladotomov) [inline fem/tmop.cpp:3067]: Is this some editor malfunction?

File: fem/tmop.cpp [hunk 3]
```diff
@@ -4324,7 +4607,11 @@ void TMOP_Integrator::ComputeNormalizationEnergies(const GridFunction &x,
          {
             if ((*surf_fit_marker)[dofs[s]] == true)
             {
-               surf_fit_gf_energy += sigma_e(s) * sigma_e(s);
+               const IntegrationPoint &ip_s = ir_s.IntPoint(s);
+               Tpr->SetIntPoint(&ip_s);
+               surf_fit_gf_energy += surf_fit_coeff->Eval(*Tpr, ip_s) *
```
Comment (vladotomov) [inline fem/tmop.cpp:4612]: I thought user-provided `Coefficients` were not part of the normalization. They would get cancelled out and have no effect.

File: fem/tmop_tools.hpp [hunk 4]
```diff
@@ -224,14 +226,26 @@ class TMOPNewtonSolver : public LBFGSSolver
    /// (ii) surface fitting weight.
    virtual void ProcessNewState(const Vector &x) const;
 
-   /** @name Methods for adaptive surface fitting weight. (Experimental) */
+   /** @name Methods for adaptive surface fitting. These control the behavior of
+       the weight and the termination of the solver. (Experimental) */
```
Comment (vladotomov) [inline fem/tmop_tools.hpp:230]: Don't you need `///@{  ... ///@}` to actually group the functions in the doxygen.

**Review by vladotomov — CHANGES_REQUESTED** [review]

File: fem/tmop.hpp [hunk 5]
```diff
@@ -217,6 +217,20 @@ class TMOP_WorstCaseUntangleOptimizer_Metric : public TMOP_QualityMetric
    virtual WorstCaseType GetWorstCaseType() { return wctype; }
 };
 
+/// Used to create TMOP_Integrators without a metric contribution.
+class TMOP_Metric_000 : public TMOP_QualityMetric
```
[the remark this answers is not in the document]
Comment (kmittal2) [reply fem/tmop.hpp:221]: Not in this PR but Mathias is using it in LiDO.

File: fem/tmop.hpp [hunk 6]
```diff
@@ -2240,6 +2253,21 @@ class TMOP_Integrator : public NonlinearFormIntegrator
...
+
+   /// Check if surface fitting is enabled
+   bool IsSurfaceFittingEnabled()
+   {
+      return surf_fit_gf != NULL || surf_fit_pos != NULL;
+   }
+
+   /// Remap the surface fitting level-set function to the provided GridFunction
+   /// Assumes @a s0 and mesh nodes share the same finite element space.
+   void ReMapSurfaceFittingLevelSet(GridFunction &s0);
```
[the remark this answers is not in the document]
Comment (kmittal2) [reply fem/tmop.hpp:2269]: Added an example of this being used in the miniapp.

**Comment by kmittal2** [comment]
After discussion with @vladotomov, I will be breaking this into smaller PRs.

**Pull request closed without merging**